### PR TITLE
60 embedded cta

### PR DESCRIPTION
**Pull request by rorJeremy**
- Add the ability to use classes with links within Contentful markdown.
- The syntax that the code is looking for is like this: `[Then let’s go!](https://www.theporch.live/awaken "Register Now"){: .button .white}`. Where the `{: .button .white}` is the syntax used specifically to add any number of classes to the url in the markdown.
- Add CustomMarkdownRender model for customizing the links that are rendered whenever it's been determined that they include a class.
- Only use the CustomMarkdownRender whenever the markdown includes links that have classes.
- Remove the `rel="nofollow"` from our url config.
- Add `target="_blank"` to any absolute urls that have classes. 

closes https://zube.io/watermarkchurch/development/c/3411

### CODE REVIEW

**Comment by rorJeremy** [comment]
Thanks for the feedback @gburgett . I added those test cases, and no doubt this feature is more resilient because of it. Some of the test cases I think are reasonable that they don't "work", like if the user completely forgets to include the link part. Others were actually very good catches on your part and required me to update the code, like handling hash locations and different amounts of spacing. I think the cases that deal with the new lines are a soft point with RedCarpet in general, so the tests are added for those scenarios are basically just serving as documentation right now to let anyone know that the link won't render as desired if you write it like that. 

Let me know if you see anything else I should address, and if not, I would like to get this merged and released so I can use it in my PR for wmresources next.

File: wcc-contentful-app/spec/dummy/app/helpers/section_helper_spec.rb [hunk 1]
```diff
@@ -50,6 +50,82 @@
...
+          helper.markdown(markdown_string)
+
+        expect(html_to_render).to include("{: .this }")
+      end
+    end
+
+    context 'When given: [links with a newline]\n(https://www.test.com){: .newline }' do
+      it 'will render the string as a mashed up mess because of the \n' do
+        markdown_string =
+          'some before text [links with a newline]\n(https://www.test.com){: .newline } and some after text'
```
Comment (gburgett) [inline wcc-contentful-app/spec/dummy/app/helpers/section_helper_spec.rb]: In Ruby strings, the single quote does not translate `\n` into an actual newline character.  Instead it's preserved as a literal `\n`.  To get it to become a newline character you should use double quotes.  I tested this with redcarpet, it does in fact recognize the following:

```ruby
irb(main):002:0> WCC::Contentful::App::SectionHelper.markdown('some before text [links with a newline]\n(https://www.test.com){: .newline } and some after text')
=> "<p>some before text [links with a newline]\\n(<a href=\"https://www.test.com)%7B:\" rel=\"nofollow\" target=\"_blank\">https://www.test.com){:</a> .newline } and some after text</p>\n"
irb(main):003:0> WCC::Contentful::App::SectionHelper.markdown("some before text [links with a newline]\n(https://www.test.com){: .newline } and some after text")
=> "<p>some before text <a href=\"https://www.test.com\" rel=\"nofollow\" target=\"_blank\">links with a newline</a>{: .newline } and some after text</p>\n"
```

File: wcc-contentful-app/spec/dummy/app/helpers/section_helper_spec.rb [hunk 2]
```diff
@@ -50,6 +50,82 @@
...
+
+        expect(html_to_render).to include(
+          "[links with a newline]\\n(<a href=\"https://www.test.com)%7B:\" target=\"_blank\">https://www.test.com){:</a> .newline }"
+        )
+      end
+    end
+    context 'When given: [newline after the parens](http://www.google.com)\n{: .test }' do
+      it 'will render the \n{: .test } after the converted hyperlink' do
+        markdown_string =
+          'some before text [newline after the parens](http://www.google.com)\n{: .test } and some after text'
```
Comment (gburgett) [inline wcc-contentful-app/spec/dummy/app/helpers/section_helper_spec.rb]: Another newline issue (see above)

**Comment by gburgett** [comment]
Hey jeremy, thanks for getting on this!  My intention was not to say that all those scenarios need to produce the link output with the intended classes.  There's room for us to train admins on how to make a button that actually renders properly.  I do however think we should be careful not to introduce any null pointer exceptions.  My primary purpose in giving all these examples is to exercise all the corners of those regexps and make sure nothing is null where it isn't supposed to be.

As long as we have a definition for what the behavior ought to be for each of those scenarios, I'm good with it!

File: wcc-contentful-app/app/helpers/wcc/contentful/app/section_helper.rb [hunk 3]
```diff
@@ -31,12 +31,16 @@ def section_id(section)
...
 
   def markdown(text)
+    markdown_links = links_within_markdown(text)
+    links_with_classes, raw_classes = gather_links_with_classes_data(markdown_links)
+
     options = {
       filter_html: true,
       hard_wrap: true,
-      link_attributes: { rel: 'nofollow', target: '_blank' },
+      link_attributes: { target: '_blank' },
```
Comment (jpowell) [inline wcc-contentful-app/app/helpers/wcc/contentful/app/section_helper.rb]: I think we need to pass whatever these are into the links that we render

File: wcc-contentful-app/app/helpers/wcc/contentful/app/section_helper.rb [hunk 4]
```diff
@@ -46,12 +50,66 @@ def markdown(text)
       tables: true
     }
 
-    renderer = ::Redcarpet::Render::HTML.new(options)
+    if links_with_classes.present?
+      renderer = ::WCC::Contentful::App::CustomMarkdownRender.new(options)
```
Comment (jpowell) [inline wcc-contentful-app/app/helpers/wcc/contentful/app/section_helper.rb]: if we properly handled `link_attributes` option, couldn't we just use this always?

Also, the pre & post processing you're doing could be done in the renderer as well? https://github.com/vmg/redcarpet#prepost-process

File: wcc-contentful-app/app/models/wcc/contentful/app/custom_markdown_render.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  class CustomMarkdownRender < Redcarpet::Render::HTML
+    def initialize(options)
+      super
+      @links_with_classes = options[:links_with_classes]
+    end
+
+    def link(link, title, content)
+      target = url_target(link)
+      if link_with_class_data =
+           @links_with_classes.find do |link_with_class|
```
Comment (jpowell) [inline wcc-contentful-app/app/models/wcc/contentful/app/custom_markdown_render.rb]: any way to make this constant time lookup?

File: wcc-contentful-app/app/models/wcc/contentful/app/custom_markdown_render.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      ActionController::Base.helpers.link_to(
+        content,
+        link,
+        hyperlink_attributes(title, link, link_class)
+      )
+    end
+
+    def hyperlink_attributes(title, url, link_class = nil)
+      default_link_attrs =
+        use_target_blank?(url) ? @options[:link_attributes] : @options[:link_attributes].except(:target)
```
Comment (jpowell) [inline wcc-contentful-app/app/models/wcc/contentful/app/custom_markdown_render.rb]: i think we should ensure `target: :_blank` is true here... other than that we're good to go

**Pull request merged**